tests/phpunit/tests/blocks/renderBlock.php

```diff
@@ -192,4 +192,123 @@ public function test_default_context_is_filterable() {
+			'tests/context-provider',
+			array(
+				'provides_context' => array( 'example' ),
+			)
+		);
+
+		register_block_type(
+			'tests/context-consumer',
+			array(
+				'uses_context'    => array( 'example' ),
+				'render_callback' => static function ( $attributes, $content, $block ) use ( &$provided_context ) {
+					$provided_context = $block->context;
+
+					return '';
+				},
+			)
+		);
+
+		// Filter the context provided by the test block.
+		add_filter(
+			'render_block_context',
+			function ( $context, $parsed_block ) {
+				if ( isset( $parsed_block['blockName'] ) && 'tests/context-provider' === $parsed_block['blockName'] ) {
+					$context['example'] = 'ok';
+				}
+
+				return $context;
+			},
+			10,
+			2
+		);
+
+		// Test inner block context when the provider block is a top-level block.
+		do_blocks(
+			<<<HTML
+<!-- wp:tests/context-provider -->
+<!-- wp:tests/context-consumer /-->
+<!-- /wp:tests/context-provider -->
+HTML
+		);
+		$this->assertTrue( isset( $provided_context['example'] ), 'Test block is top-level block: Context should include "example"' );
+		$this->assertSame( 'ok', $provided_context['example'], 'Test block is top-level block: "example" in context should be "ok"' );
+
+		// Test inner block context when the provider block is an inner block.
+		do_blocks(
+			<<<HTML
+<!-- wp:group {"layout":{"type":"constrained"}} -->
+<!-- wp:tests/context-provider -->
+<!-- wp:tests/context-consumer /-->
+<!-- /wp:tests/context-provider -->
+<!-- /wp:group -->
+HTML
+		);
+		$this->assertTrue( isset( $provided_context['example'] ), 'Test block is inner block: Block context should include "example"' );
+		$this->assertSame( 'ok', $provided_context['example'], 'Test block is inner block: "example" in context should be "ok"' );
+	}
+
+	/**
+	 * Tests that the 'render_block_context' filter arbitrary context.
+	 *
+	 * @ticket 62046
+	 */
+	public function test_render_block_context_allowed_context() {
+		$provided_context = array();
+
+		register_block_type(
+			'tests/context-consumer',
+			array(
+				'uses_context'    => array( 'example' ),
+				'render_callback' => static function ( $attributes, $content, $block ) use ( &$provided_context ) {
+					$provided_context = $block->context;
+
+					return '';
+				},
+			)
+		);
+
+		// Filter the context provided to the test block.
+		add_filter(
+			'render_block_context',
+			function ( $context, $parsed_block ) {
+				if ( isset( $parsed_block['blockName'] ) && 'tests/context-consumer' === $parsed_block['blockName'] ) {
+					$context['arbitrary'] = 'ok';
+				}
+
+				return $context;
+			},
+			10,
+			2
+		);
+
+		do_blocks(
+			<<<HTML
+<!-- wp:tests/context-consumer /-->
+HTML
+		);
+		$this->assertTrue( isset( $provided_context['arbitrary'] ), 'Test block is top-level block: Block context should include "arbitrary"' );
+		$this->assertSame( 'ok', $provided_context['arbitrary'], 'Test block is top-level block: "arbitrary" in context should be "ok"' );
```

Comment:
Per https://github.com/WordPress/wordpress-develop/pull/7522#issuecomment-2513172368, we do _not_ want arbitrary context to be passed through. The only reason this works for inner blocks is because the `$context` property is filtered instead of `$available_context`. That's not right either, but we can't easily change it now due to backward compatibility issues.

Anyway, here the correct assertion is:
```suggestion
		$this->assertFalse( isset( $provided_context['arbitrary'] ), 'Test block is top-level block: Block context should not include "arbitrary"' );
```

It also fixes the failing test.